### DISCUSSION TITLE
File Lock in Pyodide/WASM

### DISCUSSION THREAD

**Discussion by michaelweinold** [pyodide/pyodide#4150] [Q&A]
I'm trying to use the [`whoosh` package](https://github.com/mchaput/whoosh), which is a dependency of [our Brightway packages](https://github.com/brightway-lca).

To lock files, [`whoosh` uses the `fcntl` on Unix and `msvcrt` on Windows](https://github.com/mchaput/whoosh/blob/d9a3fa2a4905e7326c9623c89e6395713c189161/src/whoosh/util/filelock.py#L160C1-L163C25):

```python
if os.name == "nt":
    FileLock = MsvcrtLock
else:
    FileLock = FcntlLock
```

This means that running in in a JupyterLite environment (or on any WASM Python kernel) results in:

```
File /lib/python3.11/site-packages/whoosh/util/filelock.py:100, in FcntlLock.acquire(self, blocking)
     97 flags = os.O_CREAT | os.O_WRONLY
     98 self.fd = os.open(self.filename, flags)
--> 100 mode = fcntl.LOCK_EX
    101 if not blocking:
    102     mode |= fcntl.LOCK_NB

AttributeError: module 'fcntl' has no attribute 'LOCK_EX'
```

Since the package is pure Python, my questions are:

1. How is locking files manages in WASM?
2. How to deal with issues of this nature - add a monkeypatch (is that supported in Pyodide), install from a fork, request for changes to be made in the main branch of the package?

**Comment by rth** [comment]
It's not really a WASM issue, the file system needs to implement file locking. The in-memory file system provided by emscripten [currently doesn't](https://github.com/emscripten-core/emscripten/blob/0be5b609412d5b07f8157528df8e5088dae84858/system/lib/libc/emscripten_libc_stubs.c#L151).

There is an open emscripten [issue about this](https://github.com/emscripten-core/emscripten/issues/11797) looks like they would accept it if someone implemented it.  Maybe the situation is better with their new WASMFS file system, but last we checked it was still in development and not stable enough for us to use.

**Comment by ryanking13** [comment]
Maybe another approach would be opening a PR in `whoosh` package to replace filelock with no-op in emscripten platform, as no other process can access files simultaneously in Emscripten.

**Reply by michaelweinold** [reply]
Sounds reasonable - I'll try to open a PR with them in the next weeks.

**Reply by hoodmane** [reply]
In the meantime, what if you just monkeypatch `fcntl.LOCK_EX = 0`? Probably you'd also need `fcntl.flock` to be a no-op and `fcntl.LOCK_UN = 0`...
